CONTRIBUTING.md

```diff
@@ -15,39 +15,141 @@ Otherwise, go read Docker's
+
+Do this (one time only) to set your repo's upstream:
+
+```sh
+git remote add upstream https://github.com/docker/swarm.git
+```
+
+Do this frequently while you work to rebase your branch and bring it up to date:
+
 ```sh
-go get -d github.com/docker/swarm
+git fetch upstream master
+git rebase upstream/master
+git push origin master
 ```
 
-#### Start hacking
+Don't worry:  fetching the upstream and rebasing will not overwrite your local changes.
+
+### Build the Swarm binary
+
+**TODO:  IMPROVE THIS STEP TO ELIMINATE SYMLINK**
+
+Create a symbolic link pointing `$GOCODE/src/github.com/docker/swarm` -> `$GOCODE/src/github.com/<YOUR GITHUB USERNAME>/swarm`:
 
-First, prepare dependencies for `swarm` and try to compile it.
 ```sh
-cd src/github.com/docker/swarm
-$GOBIN/godep restore
-go test -v -race ./...
-go install
+mkdir $GOCODE/src/github.com/docker
+ln -s $GOCODE/src/github.com/<YOUR GITHUB USERNAME>/swarm $GOCODE/src/github.com/docker/swarm
 ```
 
-#### Adding New Dependencies
+Build the binary, installing it to `$GOPATH/bin/swarm`:
+
+```sh
+godep go install .
+```
+ 
+Run the binary you just created:
+
+```sh
+$GOPATH/bin/swarm help
+$GOPATH/bin/swarm manage token://<cluster_id>
+$GOPATH/bin/swarm join --addr=<node_ip:2375> token://<cluster_id>
+```
+
+#### Create a Swarm image for deployment
+
+Swarm is distributed as a Docker image with a single `swarm` binary inside.  To create the image:
+
+```sh
+docker build -t my_swarm .
+```
+
+Now you can run the same commands as above using the container:
+
+```sh
+docker run my_swarm help
+docker run -d my_swarm join --addr=<node_ip:2375> token://<cluster_id>
+docker run -d -p <manager_port>:2375 my_swarm manage token://<cluster_id>
+```
+
+For complete documentation on how to use Swarm, refer to the Swarm section of [docs.docker.com](http://docs.docker.com/).
+
+### Submit a pull request (PR)
+
+Swarm welcomes your contributions back to the project!  We follow the same contribution process and principles as other Docker projects.  This tutorial describes the [Docker contribution workflow](https://docs.docker.com/opensource/workflow/make-a-contribution/).
+
+In general, your contribution workflow for Swarm will look like this:
+
+```sh
+cd $GOCODE/src/github.com/<YOUR GITHUB USERNAME>/swarm
+git checkout -b issue-9999   # create a branch for your work
+git push origin issue-9999   # push your branch to your fork
+[change some files; git add; git commit; etc.]
+git push origin issue-9999   # re-push your branch to your fork
+```
+
+To open a PR, browse to your fork at `http://github.com/<YOUR GITHUB USERNAME>/swarm` and select the `issue-9999` branch from "Branch" drop down list.
+
+Click the green "New Pull Request" button and submit your PR.  Maintainers will review your PR and merge it or provide feedback on how to improve it.
```

Comment:
At this point, it would be helpful to add info on how to refine your PR based on comments (push to the same branch and Github magically updates the PR to add those changes)